cpp/src/arrow/array/validate.cc

```diff
@@ -622,6 +637,106 @@ struct ValidateArrayImpl {
+    if (!values_valid.ok()) {
+      return Status::Invalid("Values array invalid: ", values_valid.ToString());
+    }
+    if (data.null_count != 0) {
+      return Status::Invalid("Null count must be 0 for REE array, but was ",
+                             data.null_count);
+    }
+    if (run_ends_data.null_count != 0) {
+      return Status::Invalid("Null count must be 0 for run ends array, but was ",
+                             run_ends_data.null_count);
+    }
+    if (!run_ends_data.buffers[1]->is_cpu()) {
+      return Status::NotImplemented("Validating non-CPU run ends buffers");
+    }
+    ArraySpan span(data);
+    const RunEndsType* run_ends = ree_util::RunEnds<RunEndsType>(span);
+    if (run_ends_data.length == 0) {
+      if (data.length == 0) {
+        return Status::OK();
+      } else {
+        return Status::Invalid("REE array has non-zero length ", data.length,
+                               ", but run ends array has zero length");
+      }
+    }
+    if (run_ends[run_ends_data.length - 1] < data.offset + data.length) {
+      return Status::Invalid(
+          "Last run in run ends array ends at ", run_ends[run_ends_data.length - 1],
+          " but this array requires at least ", data.offset + data.length, " (offset ",
+          data.offset, ", length ", data.length, ")");
+    }
+    if (full_validation && run_ends_data.length != 0) {
+      const int64_t run_ends_length = ree_util::RunEndsArray(span).length;
+      int64_t last_run_end = 0;
+      int64_t physical_offset = 0;
+      int64_t physical_end = 0;
+      for (int64_t index = 0; index < run_ends_length; index++) {
+        int64_t run_end = run_ends[index];
+        if (run_end < 1) {
+          return Status::Invalid(
+              "Run ends array invalid: All run ends must be a positive integer but run "
+              "end ",
+              index, " is ", run_end);
+        }
+        if (run_end <= last_run_end) {
+          return Status::Invalid(
+              "Run ends array invalid: Each run end must be greater than the prevous "
+              "one, but run end ",
+              index, " is ", run_end, " and run end ", index - 1, " is ", last_run_end);
+        }
+        if (run_end > data.offset && last_run_end <= data.offset) {
+          physical_offset = index;
+        }
+        if (run_end >= data.offset + data.length &&
+            last_run_end < data.offset + data.length) {
+          physical_end = index + 1;
+        }
+        last_run_end = run_end;
+      }
+      ARROW_DCHECK_EQ(physical_offset, ree_util::FindPhysicalOffset(span));
+      ARROW_DCHECK_EQ(physical_end, physical_offset + ree_util::FindPhysicalLength(span));
+      if (values_data.length < physical_end) {
+        return Status::Invalid(
+            "Values array needs at least ", physical_end,
+            " elements to hold the runs described by the run ends array, but only has ",
+            values_data.length);
+      }
```

Comment:
This should not be needed in this loop. We're checking here that the run ends are monotonically increasing. We can check out of the loop that the last run end doesn't exceed the child array length.

Reply:
Agreed.